README.md
# Grav Google Analytics Plugin

The **Google Analytics** Plugin for [Grav CMS](http://github.com/getgrav/grav) allows you to integrate and configure [Google Analytics](https://www.google.com/analytics) without the need to touch any code within your Grav site.

### Features
* Choose the Google Analytics code position in the HTML document (head or body).
* Renaming of the Global (gtag) Object
* Debug Mode
* Custom Cookie Configuration. Name prefix, domain and expiration time are configurable.
* DNT header support
* Blocking IP Addresses
* Opt Out (disable tracking by the user)
* Multi-Language Support for the [Grav Administration Panel](https://github.com/getgrav/grav-plugin-admin)

## Installation

Installing the Google Analytics plugin can be done in one of two ways. The GPM (Grav Package Manager) installation method enables you to quickly and easily install the plugin with a simple terminal command, while the manual method enables you to do so via a zip file.

### GPM Installation (Preferred)

The simplest way to install this plugin is via the [Grav Package Manager (GPM)](https://learn.getgrav.org/advanced/grav-gpm) through your system's terminal (also called the command line).  From the root of your Grav install type:

    bin/gpm install ganalytics

This will install the Google Analytics plugin into your `/user/plugins` directory within Grav. Its files can be found under `/your/site/grav/user/plugins/ganalytics`.

### Manual Installation

To install this plugin, just download the zip version of this repository and unzip it under `/your/site/grav/user/plugins`. Then, rename the folder to `ganalytics`. You can find these files on [GitHub](https://github.com/escopecz/grav-ganalytics) or via [GetGrav.org](https://getgrav.org/downloads/plugins).

You should now have all the plugin files under

    /your/site/grav/user/plugins/ganalytics

> NOTE: This plugin is a modular component for Grav which requires [Grav](http://github.com/getgrav/grav) to operate.

## Configuration

Before configuring this plugin, you should copy the `user/plugins/ganalytics/ganalytics.yaml` to `user/config/plugins/ganalytics.yaml` and only edit that copy.

Here is the default configuration and an explanation of available options:

```yaml
enabled: true
trackingId: ""

position: "head"
objectName: "gtag"
dnt: true
blockedIps: []
blockedIpRanges: ["private", "loopback", "link-local"]
blockingCookie: "blockGA"

cookieConfig: false
cookiePrefix: ""
cookieDomain: ""
cookieExpires: 63072000

optOutEnabled: false
optOutMessage: "Google tracking is now disabled."

debugMode: false
```

* `enabled` Toggles if the Google Analytics plugin is turned on or off.
* `trackingId` The Google Analytics Tracking ID. This value is **required**.
_(You can also use environment variables by entering `env:VAR_NAME` as value)_
* `position` Code Position in the HTML document (`head` or `body`). Default is `head`.
* `objectName` The name for the global (gtag) object. Default is `gtag`.
* `dnt` Enables [Do Not Track header](https://www.w3.org/TR/tracking-dnt/#dnt-header-field) support. When enabled, Google Analytics code will not be embedded if a client sends DNT header. Most browsers do not send this header by default. Header support in the plugin is `enabled` by default though.
* `blockedIps` Here you can blacklist IP addresses. For those the Google Analytics script will not be embedded.
* `blockedIpRanges` Here you can blacklist IPv4 and/or IPv6 address ranges in the form `["192.177.204.1-192.177.204.254", "2001:db8::1-2001:db8::fe", ...]`. In addition to numerical ranges, the keywords "private", "loopback", "link-local" designate special IPv4 and IPv6 ranges (see RFCs 6890, 4193, 4291). For blacklisted ranges the Google Analytics script will not be embedded. By default, all three ranges are blocked. If you are using a reverse proxy that redirects traffic to the grav installation, you may need to remove "private".
* `blockingCookie` The name of a blocking cookie. When such a cookie is set, the Google Analytics script will not be embedded. Default ist `blockGA`

* `cookieConfig`: Toggles if the a custom cookie configuration should be used.
* `cookiePrefix` The cookie name prefix.
* `cookieDomain`  The cookie domain. Optional, Google default is top level domain plus one subdomain (eTLD +1). For example Grav site under https://example.com would use example.com for the cookie domain, and https://subdomain.example.com would also use example.com for the cookie domain.
* `cookieExpires` The cookie expiration time in seconds. Google default is 2 years (`63072000` seconds)

* `optOutEnabled` Toggles if opt out function is turned on or off.
* `optOutMessage` Confirmation message shown to the user when opt out function is called

* `debugMode` Toggles if Google Analytics debug mode is enabled or disabled.

## Usage

1. Sign in to your [Google Analytics account](https://analytics.google.com/).
2. Select the **Admin** tab.
3. Select an account from the dropdown in the _ACCOUNT_ column.
4. Select a property from the dropdown in the _PROPERTY_ column.
5. Under _PROPERTY_, click **Data Streams > you_data_stream_name**.
6. Copy the **MEASUREMENT ID** (a string like _G-XXXXXXXXXX_)
7. Add it to the configuration of this plugin.

To give your users the possibility to disable Google Analytics tracking you have to enable "opt out" in this plugin and put the following link somewhere in your pages, e.g. in your Privacy Declaration:

```html
<a href="javascript:gaOptout()">Disable Google Analytics</a>
```

The link must be inserted as HTML tags and not in markdown syntax. 
When this link is clicked, then the official ga-disable-cookie is set and Google stopps tracking this visitor.
For more Info about disabling the Google Analytics tracking see: https://developers.google.com/analytics/devguides/collection/gtagjs/user-opt-out

## Upgrading from 1.x

First versions of the plugin were compatible only with Universal Analytics (analytics.js). Version 2.0 onwards only supports Google Analytics 4 (gtag.js). Universal Analytics is no longer supported. This means you have to migrate your Universal Analytics properties to Google Analytics 4 data streams before performing the upgrade.

1. Sign in to your [Google Analytics account](https://analytics.google.com/) and migrate a property used by this plugin to GA4 data stream. Keep in mind that GA4 uses totally different data model and gathering techniques. Google provides [detailed guide](https://support.google.com/analytics/answer/10759417?hl=en&ref_topic=10737980) on what to expect form GA4, how it compares to UA, and what exact steps you need to take to migrate your property.
2. After migration is complete on Google side, upgrade Grav Google Analytics plugin.
3. Set new tracking ID in the plugin's configuration. It needs to match _MEASUREMENT ID_ of you GA4 data stream. Old UA tracking ID won't work anymore.
4. If you were using opt-out configuration it will be automatically reset for all your users. They will need to make opt-out decision again based on your configured policy.

After upgrade you will probably notice that a lot of old plugin configuration options are gone. That's because Google has changed how these options work in GA4. In particular:

* `anonymizeIp` and `forceSsl` options are removed because they are now enforced on Google side and cannot be disabled.
* `async` option is removed because GA4 always uses async mode.
* `cookieName` option was renamed to `cookiePrefix` because you cannot change full cookie name anymore.
* `objectName` default value was renamed from `ga` to `gtag` to keep it inline with standard GA4 snippet. If you were using custom `objectName` it won't be changed.
* `debugTrace` option was removed and `debugStatus` was renamed to `debugMode`. GA4 doesn't have debug tracing mode.

**Note: due to a bug in Grav CMS framework all these changes are not migrated automatically. At the moment these configuration options need to be removed/renamed manually.**


CHANGELOG.md
# v2.1.0
## 06/19/2023

1. [](#new)
    * Added "Do Not Track" header support which is enabled by default
2. [](#bugfix)
    * Fixed opt-out dialog output when confirmation message contains quotes
    * Fixed opt-out section title translation
    * Fixed an issue which prevented being able to install this plugin via `bin/gpm direct-install` command

# v2.0.0
## 06/09/2023

1. [](#new)
    * BREAKING CHANGE! The plugin now uses Google Analytics 4 API, which is not compatible with an old Universal Analytics functionality. You must migrate your existing Universal Analytics properties to Google Analytics 4 data streams before performing the upgrade. Please, read upgrade section in the [README](https://github.com/escopecz/grav-ganalytics/blob/master/README.md#upgrading-from-1x) file for the detailed steps.

# v1.5.2
## 03/10/2023

1. [](#bugfix)
    * Fixed changelog formating

# v1.5.1
## 03/10/2023

1. [](#bugfix)
    * Fix previously broken 1.5.0 release

# v1.5.0
## 09/22/2019

1. [](#new)
    * Added support for environment variables
    * Add capabilities to block address ranges and use a blocking cookie
    * Opt out code added
2. [](#improved)
    * Changes for General Data Protection Regulation (GDPR)

# v1.4.0
## 01/04/2017

1. [](#new)
    * Preload the Google Analytics script asynchronously
    * Choose the code position in the HTML document (head or body)
    * Custom Cookie Configuration. Name, domain and expiration time are configurable.
    * Force SSL - Send all data using SSL, even from insecure (HTTP) pages
2. [](#improved)
    * Improve plugin configuration with tab views.
    * Better use and configuration of the global object name. Please use `objectName` instead of `renameGa`.

# v1.3.0
## 12/21/2016

1. [](#new)
    * Block IP addresses (_Google Analytics code will not be embedded_)
2. [](#improved)
    * Added german translation
3. [](#bugfix)
    * Fixed the date format in the changelog 

# v1.2.0
## 08/11/2016

1. [](#new)
    * Rename the global (ga) variable of the Google Analytics object
    * Enable the debug version of the analytics.js library + Trace Debugging
      
# v1.1.0
## 08/02/2016

1. [](#new)
    * Anonymize the IP address sent to Google Analytics

# v1.0.0
## 11/08/2015

1. [](#new)
    * GA Plugin started


ganalytics.php
<?php

namespace Grav\Plugin;

use Grav\Common\Plugin;

/**
 * Class GanalyticsPlugin
 * @package Grav\Plugin
 */
class GanalyticsPlugin extends Plugin
{
    /**
     * @return array
     *
     * The getSubscribedEvents() gives the core a list of events
     * that the plugin wants to listen to. The key of each
     * array section is the event that the plugin listens to
     * and the value (in the form of an array) contains the
     * callable (or function) as well as the priority. The
     * higher the number the higher the priority.
     */
    public static function getSubscribedEvents()
    {
        return [
            'onPluginsInitialized' => ['onPluginsInitialized', 0],
            'onAssetsInitialized' => ['onAssetsInitialized', 0]
        ];
    }

    /**
     * Returns the Google Analytics cookie configuration.
     * @return array
     */
    private function getCookieConfiguration(){
        $cookie_config = $this->config->get('plugins.ganalytics.cookieConfig', false);
        if (!$cookie_config) return [];

        $cookie_config = [
          'cookie_prefix'    => $this->config->get('plugins.ganalytics.cookiePrefix', ''),
          'cookie_expires' => $this->config->get('plugins.ganalytics.cookieExpires', 63072000),
        ];

        // cookie domain
        $cookie_domain = trim($this->config->get('plugins.ganalytics.cookieDomain'));
        if (!empty($cookie_domain)) $cookie_config['cookie_domain'] = $cookie_domain;

        return $cookie_config;
    }

    /**
     * Returns the Google Analytics opt out configuration.
     * @return array
     */
    private function getOptOutConfiguration(){
        $optout_config = $this->config->get('plugins.ganalytics.optOutEnabled', false);
        if (!$optout_config) return [];

        $optout_config = [
          'optoutMessage' => addslashes(trim($this->config->get('plugins.ganalytics.optOutMessage', 'Google tracking is now disabled.'))),
          'cookieExpires' => gmdate ("D, d-M-Y H:i:s \U\T\C", $this->config->get('plugins.ganalytics.cookieExpires', 63072000) + time()),
        ];

        return $optout_config;
    }

    /**
     * Return the Google Analytics Tracking Code
     * @param string $objectName Global variable name for the GA object
     * @return string
     */
    private function getTrackingCode($objectName)
    {
        $code =
            "window.dataLayer = window.dataLayer || [];\n".
            "function {$objectName}(){dataLayer.push(arguments);}\n".
            "{$objectName}('js', new Date());\n"
        ;

        return $code;
    }

    /**
     * Return the Google Analytics Opt Out Code
     * @param string $trackingId Google Analytics Tracking ID
     * @param array $config Out Out settings
     * @return string
     */
    private function getOptOutCode($trackingId, $config)
    {
        $code = <<<JSCODE
var disableStr = 'ga-disable-$trackingId';
if (document.cookie.indexOf(disableStr + '=true') > -1) {
    window[disableStr] = true;
}
function gaOptout() {
    document.cookie = disableStr + '=true; expires={$config['cookieExpires']}; path=/';
    window[disableStr] = true;
    alert('{$config['optoutMessage']}');
}

JSCODE;
        return $code;
    }

    /**
     * Return all personalized GA settings
     * @param string $trackingId Google Analytics Tracking ID
     * @param string $objectName Global variable name for the GA object
     * @return string
     */
    private function getTrackingSettings($trackingId, $objectName)
    {
        if ($this->config->get('plugins.ganalytics.debugMode', false)) {
            $settings['debug_mode'] = true;
        } else {
            $settings = [];
        }

        $settings = array_merge($settings, $this->getCookieConfiguration());
        if (!empty($settings)) {
            try {
                $settings = json_encode($settings, JSON_THROW_ON_ERROR);
                return "{$objectName}('config', '{$trackingId}', {$settings});";
            } catch (\JsonException $e) {
                $this->grav['log']->error("plugin.{$this->name}: Invalid cookie settings - {$e->getMessage()}");
            }
        }

        return "{$objectName}('config', '{$trackingId}');";
    }

    /**
     * Do something with deprecated settings
     *
     * Due to https://github.com/getgrav/grav/issues/3697
     * this method doesn't store modified settings back to
     * YAML file. One needs to migrate deprecated configuration
     * options manually.
     */
    private function processDeprecatedSettings()
    {
        // 1.3.0 => 1.4.0
        // Field: "renameGa" => "objectName"
        $renameGa = trim($this->config->get('plugins.ganalytics.renameGa', ''));
        if (!empty($renameGa)) {
            $settings = $this->config->get('plugins.ganalytics', []);
            $settings['objectName'] = $renameGa;
            unset($settings['renameGa']);
            $this->config->set('plugins.ganalytics', $settings);
//            Plugin::saveConfig('ganalytics');
        }

        // 1.x => 2.0
        // Remove unused fields
        $fields = ['async', 'anonymizeIp', 'forceSsl', 'cookieName', 'debugTrace'];
        $settings = $this->config();
        if (array_intersect_key(array_flip($fields), $settings)) {
            unset(
              $settings['async'], $settings['anonymizeIp'], $settings['forceSsl'],
              $settings['cookieName'], $settings['debugTrace']
            );
            $this->config->set('plugins.ganalytics', $settings);
//            $this->saveConfig('ganalytics');
        }
        // Rename default "objectName" value from "ga" to "gtag"
        $objectName = $this->config->get('plugins.ganalytics.objectName');
        if ($objectName == 'ga') {
            $this->config->set('plugins.ganalytics.objectName', 'gtag');
//            Plugin::saveConfig('ganalytics');
        }
        // Rename "debugStatus" field to "debugMode"
        $debugStatus = $this->config->get('plugins.ganalytics.debugStatus');
        if (!is_null($debugStatus)) {
            $settings = $this->config->get('plugins.ganalytics', []);
            $settings['debugMode'] = $debugStatus;
            unset($settings['debugStatus']);
            $this->config->set('plugins.ganalytics', $settings);
//            Plugin::saveConfig('ganalytics');
        }
    }

    // Handle deprecated stuff when the plugin is initialized
    public function onPluginsInitialized()
    {
        $this->processDeprecatedSettings();
    }

    /**
     * Returns a packed IP address which can be directly compared to another packed IP address
     * @param string $humanReadableIPAddress IPv4 or IPv6 adress in human readable notation
     * @return string (16 byte packed representation)
     */
    private function packedIPAddress($humanReadableIPAddress)
    {
        $result = inet_pton($humanReadableIPAddress);

        if ($result == FALSE)
            return $this->packedIPAddress('::0');
        elseif (strlen($result) == 16)
            return $result;  // IPv6 native
        else
            return "\0\0\0\0\0\0\0\0\0\0\0\0" . $result;  // IPv4, expanded to IPv6 compatible length
    }

    /**
     * Returns TRUE, if a packed IP address is within the specified address range
     * @param string $packedAddress
     * @param string $range
     * @return boolean
     */
    private function inIPAdressRange($packedAddress, $range)
    {
        if ($range === 'private') {  // RFC 6890, RFC 4193
            return ($this->inIPAdressRange($packedAddress, "10.0.0.0-10.255.255.255")
                || $this->inIPAdressRange($packedAddress, "172.16.0.0-172.31.255.255")
                || $this->inIPAdressRange($packedAddress, "192.168.0.0-192.168.255.255")
                || $this->inIPAdressRange($packedAddress, "fc00::-fdff:ffff:ffff:ffff:ffff:ffff:ffff:ffff"));
        } elseif ($range === 'loopback') {  // RFC 6890
            return ($this->inIPAdressRange($packedAddress, "127.0.0.1-127.255.255.255")
                || $this->inIPAdressRange($packedAddress, "::1-::1"));
        } elseif ($range === 'link-local') {  // RFC 6890, RFC 4291
            return ($this->inIPAdressRange($packedAddress, "169.254.0.0-169.254.255.255")
                || $this->inIPAdressRange($packedAddress, "fe80::-febf:ffff:ffff:ffff:ffff:ffff:ffff:ffff"));
        } else {
            $rangeLimits = explode('-', $range);
            if (count($rangeLimits) == 2) {
                $lowerLimit = $this->packedIPAddress($rangeLimits[0]);
                $upperLimit = $this->packedIPAddress($rangeLimits[1]);
                return $lowerLimit <= $packedAddress && $packedAddress <= $upperLimit;
            }
        }

        return FALSE;
    }

    /**
     * Documents the reason for blocking GA tracking in a JavaScript comment
     * @param string $reason
     */
    private function documentBlockingReason($reason)
    {
        $this->grav['assets']->addInlineJs("/* GA tracking blocked, reason: $reason */");
    }

    /**
     * Add GA tracking JS when the assets are initialized
     */
    public function onAssetsInitialized()
    {
        // Don't proceed if we are in the admin plugin
        if ($this->isAdmin()) {
            $this->documentBlockingReason('admin plugin active');
            return;
        }

        // Don't proceed if there is no GA Tracking ID
        $trackingId = trim($this->config->get('plugins.ganalytics.trackingId', ''));

        // Add support for environment variables:
        if (preg_match('/env:(.*)/', $trackingId, $match)) {
            $trackingId = getenv($match[1]);
        }

        if (empty($trackingId)) {
            $this->documentBlockingReason('trackingId not configured');
            return;
        }

        $dntRespected = $this->config->get('plugins.ganalytics.dnt', true);
        if ($dntRespected && !empty($_SERVER['HTTP_DNT'])) {
            $this->documentBlockingReason("Do Not Track header sent by the client");
            return;
        }

        // Don't proceed if a blocking cookie is set
        $blockingCookieName = $this->config->get('plugins.ganalytics.blockingCookie', '');
        if (!empty($blockingCookieName) && !empty($_COOKIE[$blockingCookieName])) {
            $this->documentBlockingReason("blocking cookie \"$blockingCookieName\" is set");
            return;
        }

        // Don't proceed if the IP address is blocked
        $blockedIps = $this->config->get('plugins.ganalytics.blockedIps', []);
        if (in_array($_SERVER['REMOTE_ADDR'], $blockedIps)) {
            $this->documentBlockingReason("client ip " . $_SERVER['REMOTE_ADDR'] . " is in blockedIps");
            return;
        }

        // Don't proceed if the IP address is within a blocked range
        $packedClientIpAddress = $this->packedIPAddress($_SERVER['REMOTE_ADDR']);
        $blockedIpRanges = $this->config->get('plugins.ganalytics.blockedIpRanges', []);
        foreach ($blockedIpRanges as $blockedIpRange) {
            if ($this->inIPAdressRange($packedClientIpAddress, $blockedIpRange)) {
                $this->documentBlockingReason("client ip " . $_SERVER['REMOTE_ADDR'] . " is in range \"" . $blockedIpRange . "\"");
                return;
            }
        }

        // Parameters
        $objectName = trim($this->config->get('plugins.ganalytics.objectName', 'gtag'));
        $position   = trim($this->config->get('plugins.ganalytics.position', 'head'));

        // Opt Out and Tracking Code and settings
        $code = ''; // init
        $optout_config = $this->getOptOutConfiguration();
        if (!empty($optout_config)) {
            $code .= $this->getOptOutCode($trackingId, $optout_config);
        }
        $settings = $this->getTrackingSettings($trackingId, $objectName);
        $code .= $this->getTrackingCode($objectName);
        $code .= $settings;

        // Embed Google Analytics script
        $group = ($position == 'body') ? 'bottom' : null;

        $this->grav['assets']->addJs("https://www.googletagmanager.com/gtag/js?id={$trackingId}", 9, true, 'async', $group);
        $this->grav['assets']->addInlineJs($code, null, $group);
    }
}
